src/Microsoft.Azure.WebJobs.Host/Bindings/QueueBindingProvider.cs

```diff
@@ -0,0 +1,262 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Host.Executors;
+using Microsoft.Azure.WebJobs.Host.Protocols;
+using Microsoft.Azure.WebJobs.Host.Queues;
+using Microsoft.Azure.WebJobs.Host.Storage;
+using Microsoft.Azure.WebJobs.Host.Storage.Queue;
+using Microsoft.WindowsAzure.Storage.Queue;
+
+namespace Microsoft.Azure.WebJobs.Host.Bindings
+{
+    // Write up bindinging rules for [Queue] attribute. 
+    // This is fundemantentally an IAsyncCollector<IStorageQueueMessage>
+    internal class QueueBindingProvider
+    {
+        // Fields that the various binding funcs need to close over. 
+        private readonly IStorageAccountProvider _accountProvider;
+        private readonly IContextGetter<IMessageEnqueuedWatcher> _messageEnqueuedWatcherGetter;
+
+        // Use the static Build method to create. 
+        // Constructor is just for capturing instance fields used in func closures. 
+        private QueueBindingProvider(
+               IStorageAccountProvider accountProvider,
+               IContextGetter<IMessageEnqueuedWatcher> messageEnqueuedWatcherGetter)
+        {
+            _accountProvider = accountProvider;
+            _messageEnqueuedWatcherGetter = messageEnqueuedWatcherGetter;
+        }
+
+        public static IBindingProvider Build(
+            IStorageAccountProvider accountProvider,
+            IContextGetter<IMessageEnqueuedWatcher> messageEnqueuedWatcherGetter,
+            INameResolver nameResolver,
+            IConverterManager converterManager)
+        {
+            var closure = new QueueBindingProvider(accountProvider, messageEnqueuedWatcherGetter);
```

Comment:
As an outsider, this closure naming for a new binding provider is somewhat confusing :)